File: BalancedHeightBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:

    #Here we traverse all the nodes. For each node we get its left height and right height. If the difference b/w the heights is more than 1 then we make the flag false. for each node we return the max of its left or right height and plus 1(self)
    #Time complexity - O(n)
    #Space complexity - O(1)
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.flag = True
        
        def helper(root):
            if not root:
                return 0
            
            left = helper(root.left)
            right = helper(root.right)

            if abs(left-right)>1:
                self.flag = False

            return 1+max(left,right)
        
        helper(root)
        return self.flag

File: BalancedHeightBinaryTree.java
// Time Complexity : O(n) //N being total nodes of Tree
// Space Complexity : O(h) // Height of Tree. These many stacks calls are being stored in worst case
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : Came up with logic. To codify, it took a lot of time. Condition checks - ufff!

/*
Perform Inorder traversal of Tree
Upon base condition hit, return 0. Then for each node count right and left height.
Take min of them while they return to their node, and add 1 - That is for parent node itself.
At any place, left and right height difference is greater than 1 then mark as false.

*/

public class BalancedHeightBinaryTree {

      public class TreeNode {
          int val;
          TreeNode left;
          TreeNode right;
          TreeNode() {}
          TreeNode(int val) { this.val = val; }
          TreeNode(int val, TreeNode left, TreeNode right) {
              this.val = val;
              this.left = left;
              this.right = right;
          }
      }

        boolean flag = true;
        public boolean isBalanced(TreeNode root) {
            helper(root); //here the function is returning integer, but main is boolean
            return flag;
        }

        //recursive function
        public int helper(TreeNode root){
            //base
            if(root == null) return 0;

            //logic
            int left = helper(root.left);
            int right = helper(root.right);

            if( Math.abs(left-right) > 1 ) {
                flag = false;
            }
            return Math.max(left, right) + 1; //+1 is for parent node itself when that returns to its parent.
        }
    }